Index: libdrm/xf86drmMode.c
===================================================================
--- libdrm.orig/xf86drmMode.c	2010-02-01 12:41:44.603591443 +1100
+++ libdrm/xf86drmMode.c	2010-02-01 12:46:55.943922337 +1100
@@ -620,6 +620,28 @@
 			return 0;
 	}
 
+	/* Ubuntu's linux-backports-modules renames the drm module to drm_lbm
+	   to avoid conflicting with existing drm modules.  We therefore need
+	   to search in $DIR/drm_lbm as well as $DIR/drm */
+	sprintf(pci_dev_dir, "/sys/bus/pci/devices/%04x:%02x:%02x.%d/drm_lbm",
+		domain, bus, dev, func);
+
+	sysdir = opendir(pci_dev_dir);
+	if (sysdir) {
+		dent = readdir(sysdir);
+		while (dent) {
+			if (!strncmp(dent->d_name, "controlD", 8)) {
+				found = 1;
+				break;
+			}
+
+			dent = readdir(sysdir);
+		}
+		closedir(sysdir);
+		if (found)
+			return 0;
+	}
+
 	sprintf(pci_dev_dir, "/sys/bus/pci/devices/%04x:%02x:%02x.%d/",
 		domain, bus, dev, func);
 
