diff -rup libdrm-2.4.14~/libdrm/xf86drm.c libdrm-2.4.14/libdrm/xf86drm.c
--- libdrm-2.4.14~/libdrm/xf86drm.c	2009-07-10 04:17:53.000000000 +0100
+++ libdrm-2.4.14/libdrm/xf86drm.c	2009-12-11 04:22:07.094463655 +0000
@@ -2498,7 +2498,6 @@ int drmSetMaster(int fd)
 {
 	int ret;
 
-	fprintf(stderr,"Setting master \n");
 	ret = ioctl(fd, DRM_IOCTL_SET_MASTER, 0);
 	return ret;
 }
@@ -2506,7 +2505,7 @@ int drmSetMaster(int fd)
 int drmDropMaster(int fd)
 {
 	int ret;
-	fprintf(stderr,"Dropping master \n");
+
 	ret = ioctl(fd, DRM_IOCTL_DROP_MASTER, 0);
 	return ret;
 }
