Index: libdrm/xf86drm.c
===================================================================
--- libdrm.orig/xf86drm.c	2009-12-15 20:35:52.000000000 +0200
+++ libdrm/xf86drm.c	2009-12-15 21:13:17.000000000 +0200
@@ -2494,7 +2494,6 @@
 {
 	int ret;
 
-	fprintf(stderr,"Setting master \n");
 	ret = ioctl(fd, DRM_IOCTL_SET_MASTER, 0);
 	return ret;
 }
@@ -2502,7 +2501,7 @@
 int drmDropMaster(int fd)
 {
 	int ret;
-	fprintf(stderr,"Dropping master \n");
+
 	ret = ioctl(fd, DRM_IOCTL_DROP_MASTER, 0);
 	return ret;
 }
