Index: b/nouveau/nouveau_bo.c
===================================================================
--- a/nouveau/nouveau_bo.c
+++ b/nouveau/nouveau_bo.c
@@ -198,14 +198,14 @@
 	if (flags & (NOUVEAU_BO_VRAM | NOUVEAU_BO_GART)) {
 		ret = nouveau_bo_kalloc(nvbo, NULL);
 		if (ret) {
-			nouveau_bo_ref(NULL, (void *)nvbo);
+			nouveau_bo_ref(NULL, (void *)&nvbo);
 			return ret;
 		}
 
 		if (flags & NOUVEAU_BO_PIN) {
 			ret = nouveau_bo_pin((void *)nvbo, nvbo->flags);
 			if (ret) {
-				nouveau_bo_ref(NULL, (void *)nvbo);
+				nouveau_bo_ref(NULL, (void *)&nvbo);
 				return ret;
 			}
 		}
