commit c5c503b519777fb9ebbaa22681a1ed81717c1d71
Author: Marcin Slusarz <marcin.slusarz@gmail.com>
Date:   Wed Dec 30 21:32:40 2009 +0100

    nouveau: add BEGIN_RING_NI

diff --git a/nouveau/nouveau_pushbuf.h b/nouveau/nouveau_pushbuf.h
index c7ac8c4..46982af 100644
--- a/nouveau/nouveau_pushbuf.h
+++ b/nouveau/nouveau_pushbuf.h
@@ -111,6 +111,14 @@ BEGIN_RING(struct nouveau_channel *chan, struct nouveau_grobj *gr,
 	chan->pushbuf->remaining -= (size + 1);
 }
 
+/* non-incrementing BEGIN_RING */
+static __inline__ void
+BEGIN_RING_NI(struct nouveau_channel *chan, struct nouveau_grobj *gr,
+	   unsigned mthd, unsigned size)
+{
+	BEGIN_RING(chan, gr, mthd | 0x40000000, size);
+}
+
 static __inline__ void
 FIRE_RING(struct nouveau_channel *chan)
 {
